compiler/src/dotty/tools/dotc/transform/TypeUtils.scala

```diff
@@ -27,5 +27,28 @@ object TypeUtils {
       case self: MethodicType => self
       case _ => if (ctx.erasedTypes) MethodType(Nil, self) else ExprType(self)
     }
+
+    /** The arity of this tuple type, which can be made up of Unit, TupleX and `*:` pairs,
+     *  or -1 if this is not a tuple type.
+     */
+    def tupleArity(implicit ctx: Context): Int = self match {
```

Comment:
Could be `tailtec`:
```scala
def tupleArity(implicit ctx: Context): Int = {
  def tupleArity(tp: Type, arity: Int): Int = tp match {
    case AppliedType(tycon, _ :: tl :: Nil) if tycon.isRef(defn.PairClass) =>
      tupleArity(tl, arity + 1)
   case tp1 =>
      if (tp1.isRef(defn.UnitClass)) arity
      else if (defn.isTupleClass(tp1.classSymbol)) arity + tp1.dealias.argInfos.length
      else -1
  }
  tupleArity(self, 0)
}
```

Reply:
I believe that version behaves differently for *: sequences that do not end in `Unit`.

Reply:
Maybe, still there exists a correct tailrec version of the method. Can be revisited later.